src/DeterministicNoiseSimulator.cpp

```diff
@@ -106,169 +122,214 @@ std::map<std::string, double> DeterministicNoiseSimulator::DeterministicSimulate
-            return qc::MatrixDD::terminal(tmp.w);
+        if (originalEdge.isTerminal()) {
+            return dEdge::terminal(tmp.w);
         } else {
-            tmp.p = density_op.p;
+            tmp.p = originalEdge.p;
             return tmp;
         }
     }
-
-    std::vector used_qubits = op->getTargets();
-    for (auto control: op->getControls()) {
-        used_qubits.push_back(control.qubit);
-    }
-    sort(used_qubits.begin(), used_qubits.end());
-
-    std::array<qc::MatrixDD, 4> new_edges{};
-    for (int i = 0; i < 4; i++) {
-        if (density_op.p->e[i].w.approximatelyZero()) {
-            // applyNoiseEffects2 returns a pointer to the terminal node with edge weight 0
-            new_edges[i] = qc::MatrixDD::terminal(density_op.p->e[i].w);
-            continue;
-        }
-
-        // Check if the target of the current edge is in the Compute table. Note that I check for the target node of
-        // the current edge if noise needs to be applied or not
-        qc::MatrixDD tmp = noiseLookup(density_op.p->e[i], used_qubits);
-
-        if (tmp.p != nullptr) {
-            new_edges[i] = tmp;
-            continue;
+    auto originalCopy = originalEdge;
+    originalCopy.w    = dd::Complex::one;
+
+    // Check if the target of the current edge is in the "compute table".
+
+    //    auto noiseLookUpResult = dd->densityNoise.lookup(originalCopy, used_qubits);
+    //
+    //    if (noiseLookUpResult.p != nullptr) {
+    //        auto tmpComplexValue = dd->cn.getCached();
+    //        CN::mul(tmpComplexValue, noiseLookUpResult.w, originalEdge.w);
+    //        noiseLookUpResult.w = dd->cn.lookup(tmpComplexValue);
+    //        dd->cn.returnToCache(tmpComplexValue);
+    //        return noiseLookUpResult;
+    //    }
+
+    std::array<dEdge, 4> new_edges{};
+    for (short i = 0; i < 4; i++) {
+        if (firstPathEdge || i == 1) {
+            // If I am to the useDensityMatrix I cannot minimize the necessary operations anymore
+            dEdge::applyDmChangesToEdges(&originalCopy.p->e[i], nullptr);
+            new_edges[i] = applyNoiseEffects(originalCopy.p->e[i], used_qubits, true);
+            dEdge::revertDmChangesToEdges(&originalCopy.p->e[i], nullptr);
+        } else if (i == 2) {
+            // Size e[1] == e[2] (due to density matrix representation), I can skip calculating e[2]
+            new_edges[2].p = new_edges[1].p;
+            new_edges[2].w = new_edges[1].w;
+        } else {
+            dEdge::applyDmChangesToEdges(&originalCopy.p->e[i], nullptr);
+            new_edges[i] = applyNoiseEffects(originalCopy.p->e[i], used_qubits, false);
+            dEdge::revertDmChangesToEdges(&originalCopy.p->e[i], nullptr);
         }
-
-        new_edges[i] = ApplyNoiseEffects(density_op.p->e[i], op, maxDepth);
-
-        // Adding the operation to the operation table
-        noiseInsert(density_op.p->e[i], used_qubits, new_edges[i]);
     }
+    dd::Edge<dd::dNode> e = {};
```

Comment:
why is this type suddenly used here? try to use consistent types throughout, i.e., `dd::dEdge` in this case.

Reply:
done